File: backend/app/models/checklist.py
from datetime import datetime
from sqlalchemy import Column, Integer, String, DateTime, Date, Boolean, ForeignKey
from sqlalchemy.orm import relationship

from app.database import Base


class EndOfDayChecklist(Base):
    __tablename__ = "end_of_day_checklists"

    id = Column(Integer, primary_key=True, autoincrement=True)
    user_id = Column(Integer, nullable=False)

    checklist_date = Column(Date, nullable=False)

    is_completed = Column(Boolean, default=False)
    completed_at = Column(DateTime)

    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)

    items = relationship("ChecklistItem", back_populates="checklist", cascade="all, delete-orphan")


class ChecklistItem(Base):
    __tablename__ = "checklist_items"

    id = Column(Integer, primary_key=True, autoincrement=True)
    checklist_id = Column(Integer, ForeignKey("end_of_day_checklists.id", ondelete="CASCADE"), nullable=False)
    schedule_item_id = Column(Integer, ForeignKey("schedule_items.id", ondelete="SET NULL"))

    title = Column(String(200), nullable=False)
    is_checked = Column(Boolean, default=False)
    checked_at = Column(DateTime)

    checklist = relationship("EndOfDayChecklist", back_populates="items")

File: mobile/lib/models/checklist.dart
class EndOfDayChecklist {
  final int id;
  final int userId;
  final DateTime checklistDate;
  final bool isCompleted;
  final DateTime? completedAt;
  final List<ChecklistItem> items;

  EndOfDayChecklist({
    required this.id,
    required this.userId,
    required this.checklistDate,
    required this.isCompleted,
    this.completedAt,
    required this.items,
  });

  factory EndOfDayChecklist.fromJson(Map<String, dynamic> json) =>
      EndOfDayChecklist(
        id: json['id'] as int,
        userId: json['user_id'] as int,
        checklistDate: DateTime.parse(json['checklist_date'] as String),
        isCompleted: json['is_completed'] as bool? ?? false,
        completedAt: json['completed_at'] != null
            ? DateTime.parse(json['completed_at'] as String)
            : null,
        items: (json['items'] as List<dynamic>?)
                ?.map(
                    (e) => ChecklistItem.fromJson(e as Map<String, dynamic>))
                .toList() ??
            [],
      );
}

class ChecklistItem {
  final int id;
  final int checklistId;
  final int? scheduleItemId;
  final String title;
  bool isChecked;
  final DateTime? checkedAt;

  ChecklistItem({
    required this.id,
    required this.checklistId,
    this.scheduleItemId,
    required this.title,
    required this.isChecked,
    this.checkedAt,
  });

  factory ChecklistItem.fromJson(Map<String, dynamic> json) => ChecklistItem(
    id: json['id'] as int,
    checklistId: json['checklist_id'] as int,
    scheduleItemId: json['schedule_item_id'] as int?,
    title: json['title'] as String,
    isChecked: json['is_checked'] as bool? ?? false,
    checkedAt: json['checked_at'] != null
        ? DateTime.parse(json['checked_at'] as String)
        : null,
  );
}
